arbor/util/ratelem.hpp

```diff
@@ -73,29 +72,49 @@ double rat_eval(const std::array<double, 1+a+c>& g, const std::array<double, 2+a
             h[i] = p[i+1] + k/((x - i)/(g[i+1]-p[i+1]) + (i+k - x)/(g[i]-p[i+1]));
         }
 
-        return rat_eval<0, c-1, k+1, upper>(h, g, x);
+        return recursive_rat_eval<0, c-1, k+1, upper>(h, g, x);
     }
     else {
         std::array<double, a+c> h;
         rat_interpolate<k, upper>(h, g, x);
-        return rat_eval<a-1, c, k+1, upper>(h, g, x);
+        return recursive_rat_eval<a-1, c, k+1, upper>(h, g, x);
     }
 }
 
+template <unsigned p, unsigned q>
+double rat_eval(const std::array<double, 1+p+q>& data, double x) {
+    // upper => interpolate polynomials first;
+    // !upper => interpolate reciprocal polynomials first;
+    // a is the number of (reciprocal) polynomial interpolation steps;
+    // c is the number of 'rhombus' interpolation steps.
+
+    constexpr bool upper = p>=q;
+    constexpr unsigned a = upper? p-q+(q>0): q-p+(p>0);
+    constexpr unsigned c = p+q-a;
 
-template <unsigned a, unsigned c, unsigned k, bool upper>
-double rat_eval(const std::array<double, 1+a+c>& g, double x) {
     if constexpr (a==0 && c==0) {
-        return g[0];
+        return data[0];
+    }
+    else if constexpr (a==1 && c==1) {
+        // Special case when p==1 and q==1 so that we can have
+        // well-defined behaviour in the case where the interpolants
+        // are strictly monotonic, but might take non-finite values.
+        // (Used for ixa interpolation in embed_pwlin.cpp).
+
+        double y0 = data[0];
+        double y1 = data[1];
+        double y2 = data[2];
+
+        return y1+(2*x-1)/(x/(y2-y1)+(1-x)/(y1-y0));
```

Comment:
It can be derived from the formula you get from the generic case, cancelling some terms. Or more directly, it's an order (1, 1) rational function f(x) (i.e. p(x)/q(x) where p and q are order 1 polynomials), and by inspection satisfies f(0) = y0, f(0.5) = y1, f(1) = y2, and so must be the desired interpolating function.